Fix broadcast for 0.5

davidagold opened the pull request:
This PR transitions the NullableArrays `broadcast` interface to incorporate that present in base: https://github.com/JuliaLang/julia/blob/913b24de9c68b6c79a767bc2f6caacea249b3198/base/broadcast.jl. Functionality remains the same. I made one minor revision to the tests, which originally checked (for some reason) that `broadcast!` and `map!` give identical behavior when fed only a single array argument. They don't -- see https://github.com/JuliaLang/julia/issues/12277. The fact that those tests were ever not problematic means that the semantics of `broadcast` specialized for `NullableArray` arguments were not consistent with those in base. Now they are.

quinnj commented:
Hmmm, looks like this breaks 0.4. We also still have a few other fixes needed for 0.5 (which is why that check failed).

davidagold commented:
The push failed because I hadn't pulled Milan's fixes for the other stuff. The PR failed because I forgot I couldn't just nix all the old definitions >_> Fixing that now.

davidagold commented:
"Rebasing" this in a crude yet (hopefully effective) manner.

davidagold committed:
Fix broadcast for 0.5

Transition to broadcast interface in
https://github.com/JuliaLang/julia/commit/522547ba3806bd2f0aae24f850f96607c7479000

davidagold commented:
Comments/suggestions @quinnj @nalimilan ?

nalimilan commented on src/broadcast.jl:
```diff
@@ -1,64 +1,141 @@
+        nullcheck
+    end
+
+    @generated function Base.Broadcast._broadcast!{M,XT,nargs}(f, Z::NullableArray, indexmaps::M, Xs::XT, ::Type{Val{nargs}}; lift=false)
```
Cut line at 92 chars.

nalimilan commented on src/broadcast.jl:
```diff
@@ -71,19 +148,17 @@ end
-        func(B, As...)
-        return B
+    @inline function Base.broadcast!(f, Z::NullableArray, Xs::NullableArray...; lift=false)
+    # @inline function Base.broadcast!{nargs}(f, Z::NullableArray, Xs::Vararg{NullableArray, nargs})
```
Why keep this commented out line and the one below?

nalimilan commented on src/broadcast.jl:
```diff
@@ -96,12 +171,12 @@ disabled by default. Note that this method's signature specifies the source
+                      Xs...; lift=lift)
 end
-
+#
```
?

nalimilan commented on test/broadcast.jl:
```diff
@@ -72,12 +72,12 @@ module TestBroadcast
     )
-        map!(g, array)
-        broadcast!(f, nullablearray)
+        map(g, array)
```
I guess there's a good reason to remove this, but then will `broadcast!` be tested? (At the very least, you should adapt the comment above).

nalimilan commented on src/broadcast.jl:
```diff
@@ -1,64 +1,141 @@
+        This method implements the same behavior as that of `broadcast!` when called on
+        regular `Array` arguments. It also includes the `lift` keyword argument, which
+        when set to true will lift `f` over the entries of the `As`. Lifting is
+        disabled by default. Note that this method's signature specifies the destination
```
Maybe cut into two paragraphs here?

nalimilan replied:
Should also do the same the corresponding >= 0.5 block.

nalimilan commented on src/broadcast.jl:
```diff
@@ -71,19 +148,17 @@ end
+        check_broadcast_shape(sz, Xs...)
+        mapindex = map(x->newindexer(sz, x), Xs)
+        # println("mapindex = ", mapindex)
+        Base.Broadcast._broadcast!(f, Z, mapindex, Xs, Val{nargs}; lift=lift)
```
You don't need `Base.Broadcast.` since you called `using`.

nalimilan commented:
Cool. I could only make superficial comments. Please merge and tag a new release as soon as you can!

quinnj commented:
Travis is happy :)

davidagold commented:
@quinnj @nalimilan I don't know what to make of the push failure. I can't recreate it locally. Possibly just some floating point weirdness? Though I don't see how that would be possible...

EDIT: It's especially weird, since the push and pr versions should be identical.

quinnj commented:
I restarted the build. I'll also pull locally to see if I can reproduce.

quinnj commented:
Must have been intermittent; it all looks good now.

quinnj commented:
Hmmm.....any idea what the map failure is, @davidagold? https://travis-ci.org/JuliaStats/NullableArrays.jl/jobs/139840111

davidagold commented:
... No? When did that happen?

Edit: master seems fine locally...
